src/runtime/on/index.ts

```diff
@@ -0,0 +1,133 @@
+import { MaybePromise, SideEffector } from '../../lib/utils'
+
+export type LazyState = {
+  callbacks: {
+    runtime: {
+      before: SideEffector[]
+      after: SideEffector[]
+      schema: {
+        before: SideEffector[]
+        after: SideEffector[]
+      }
+    }
+  }
+}
+
+function createLazyState(): LazyState {
+  return {
+    callbacks: {
+      runtime: {
+        before: [],
+        after: [],
+        schema: {
+          before: [],
+          after: [],
+        },
+      },
+    },
+  }
+}
+
+/**
+ * Public component interface
+ */
+export interface On {
```

Comment:
Sigh, I still cannot get behind this API. Realized it would require another level of nesting to handle distinction between `.start` / `.stop`.

I have another idea I'm going to try...